Feature/XCTAssert+URLRequest

gentique opened the pull request:
Testing and asserting URLRequest-s isn't very reliable. Firstly there's no clear indicator as to why two URLRequests are different, it gives a generic message when it fails: 
<img width="505" alt="Screen Shot 2022-09-09 at 3 22 32 PM" src="https://user-images.githubusercontent.com/15278390/189359775-e02b227d-f57b-4148-8111-88b4266d0eaa.png">

--

The second one is that it does not compare the `httpBody` and therefore can give you a passing test when reality it isn't.
<img width="482" alt="Screen Shot 2022-09-09 at 3 13 41 PM" src="https://user-images.githubusercontent.com/15278390/189359154-3c088049-39c7-4d0f-9030-793f98f20c4b.png">

--

If we actually test for the `httpBody` we can see a failing test:
<img width="684" alt="Screen Shot 2022-09-09 at 3 14 21 PM" src="https://user-images.githubusercontent.com/15278390/189359249-9f8b7161-689e-494f-bb57-9125b4efbf1f.png">

--

With the help of this extension we can find out specifically what is failing and therefore fix it quickly.
<img width="798" alt="Screen Shot 2022-09-09 at 3 15 46 PM" src="https://user-images.githubusercontent.com/15278390/189359366-63a596e5-f009-41f2-8e53-a4fdc4b37ac8.png">

tonikocjan commented on Sources/Core/Extensions/Other/XCTestCase+URLRequest.swift:
```diff
@@ -0,0 +1,19 @@
+import XCTest
+
+public extension XCTestCase {
+  func XCTAssertEqualURLRequests(_ lhs: URLRequest?, _ rhs: URLRequest?, file: StaticString = #filePath, line: UInt = #line) {
```
Maybe we can override the existing ‘XCAssertEqual’ instead? I would also rather see non-optional arguments.

gentique replied:
> Maybe we can override the existing ‘XCAssertEqual’ instead? I would also rather see non-optional arguments.

Hey Toni, Thanks for the feedback.

I tried overloading the method, but I'm getting some errors. Could you give me an example of how it would look like? It would greatly help me out.

Regarding the optional arguments, I have two points that I would like to mention:

The first one would be that sending optional values is already supported by XCTest methods, so it wouldn't fall outside their practices, and the second one is more related to how this method came to life.
The particular use case was for testing a URLSession client. Since it accepts URLRequest, I had to intercept the request and assert that it was the correct one. By that flow, if the implementation is wrong, the URLRequest can be optional. You can argue that you can assert unwrap the optional value, but I deemed it unnecessary since you can already XCTAssert optional and non-optional values.

tonikocjan replied:
Let's keep optional arguments, I agree.

> Could you give me an example of how it would look like

This compiles for me:

```swift
func XCTAssertEqual(_ lhs: URLRequest?, _ rhs: URLRequest?, file: StaticString = #filePath, line: UInt = #line) {
  XCTAssertTrue(lhs == rhs)
  guard let lhs = lhs, let rhs = rhs else { return }
  XCTAssertEqual(lhs.url, rhs.url, "url", file: file, line: line)
  XCTAssertEqual(lhs.httpMethod, rhs.httpMethod, "httpMethod", file: file, line: line)
  XCTAssertEqual(lhs.httpBody, rhs.httpBody, "httpBody", file: file, line: line)
  XCTAssertEqual(lhs.allHTTPHeaderFields, rhs.allHTTPHeaderFields, "allHTTPHeaderFields", file: file, line: line)
}
```

gentique replied:
If I try to add it as an extension to XCTestCase then it gives errors such as: 
`Use of 'XCTAssertEqual' refers to instance method rather than global function 'XCTAssertEqual(_:_:_:file:line:)' in module 'XCTest'`

tonikocjan replied:
Just make it a global function (other `XCTest...` methods are global as well).

tonikocjan commented on Sources/Core/Extensions/Other/XCTestCase+URLRequest.swift:
```diff
@@ -0,0 +1,19 @@
+    guard let lhs = lhs, let rhs = rhs else { return }
+    XCTAssertEqual(lhs.url, rhs.url, "url", file: file, line: line)
+    XCTAssertEqual(lhs.httpMethod, rhs.httpMethod, "httpMethod", file: file, line: line)
+    XCTAssertEqual(lhs.httpBody ?? Data(), rhs.httpBody ?? Data(), "httpBody", file: file, line: line)
+    XCTAssertEqual(lhs.allHTTPHeaderFields ?? [:], rhs.allHTTPHeaderFields ?? [:], "allHTTPHeaderFields", file: file, line: line)  }
```
You can omit the default values (`Data()` and `[:]`).

gentique replied:
The reason why I defaulted to Empty Data and Dictionary is that it happens that URLRequest sets nil values as empty, and this can mess up with testing.
<img width="637" alt="Screen Shot 2022-09-12 at 2 44 22 PM" src="https://user-images.githubusercontent.com/15278390/189656924-ee188a4a-1998-42a5-8e71-8e4ceff19682.png">

tonikocjan commented on Sources/Core/Extensions/Other/XCTestCase+URLRequest.swift:
```diff
@@ -0,0 +1,19 @@
+import XCTest
+
+public extension XCTestCase {
+  func XCTAssertEqualURLRequests(_ lhs: URLRequest?, _ rhs: URLRequest?, file: StaticString = #filePath, line: UInt = #line) {
```
Should we also convert arguments to `@autoclosure`, to follow the XCTest design?

gentique replied:
I tested this out as follows:
```swift
func XCTAssertEqual(_ lhs: @autoclosure() -> URLRequest?, _ rhs: @autoclosure() -> URLRequest?, file: StaticString = #filePath, line: UInt = #line) {
  XCTAssertTrue(lhs() == rhs(), file: file, line: line)
  guard let lhs = lhs(), let rhs = rhs() else { return }
  XCTAssertEqual(lhs.url, rhs.url, "url", file: file, line: line)
  XCTAssertEqual(lhs.httpMethod, rhs.httpMethod, "httpMethod", file: file, line: line)
  XCTAssertEqual(lhs.httpBody ?? Data(), rhs.httpBody ?? Data(), "httpBody", file: file, line: line)
  XCTAssertEqual(lhs.allHTTPHeaderFields ?? [:], rhs.allHTTPHeaderFields ?? [:], "allHTTPHeaderFields", file: file, line: line)
}
```
But since the original XCTAssertEqual is of generic type I think it's being prioritized instead of our function.

borut-t commented on Sources/Core/Extensions/Other/XCTestCase+URLRequest.swift:
```diff
@@ -0,0 +1,19 @@
+import XCTest
+
+public extension XCTestCase {
+  func XCTAssertEqualURLRequests(_ lhs: URLRequest?, _ rhs: URLRequest?, file: StaticString = #filePath, line: UInt = #line) {
```
I would rename to singularity `XCTAssertEqualURLRequest`.

borut-t reviewed (approved):
Great 👌

borut-t commented on Sources/Core/Extensions/Other/XCTestCase+URLRequest.swift:
```diff
@@ -0,0 +1,20 @@
+import XCTest
+
+public extension XCTestCase {
+  func XCTAssertEqualURLRequest(_ lhs: URLRequest?, _ rhs: URLRequest?, file: StaticString = #filePath, line: UInt = #line) {
```
Let's put this under `XCTestCase+PovioKit` file instead.

borut-t commented on Sources/Core/Extensions/Other/XCTestCase+PovioKit.swift:
```diff
@@ -0,0 +1,20 @@
+import XCTest
+
+public extension XCTestCase {
+  func XCTAssertEqualURLRequest(_ lhs: URLRequest?, _ rhs: URLRequest?, file: StaticString = #filePath, line: UInt = #line) {
+    XCTAssertEqual(lhs, rhs, file: file, line: line)
+    guard let lhs = lhs, let rhs = rhs else { return }
+    XCTAssertEqual(lhs.url, rhs.url, "url", file: file, line: line)
+    XCTAssertEqual(lhs.httpMethod, rhs.httpMethod, "httpMethod", file: file, line: line)
+    XCTAssertEqual(lhs.httpBody ?? Data(), rhs.httpBody ?? Data(), "httpBody", file: file, line: line)
+    XCTAssertEqual(lhs.allHTTPHeaderFields ?? [:], rhs.allHTTPHeaderFields ?? [:], "allHTTPHeaderFields", file: file, line: line)
+  }
```
Can we make tests for this method as well?

gentique commented:
@borut-t Please check my latest commits, I added the same file as you did on #199, this way you can easily rebase and add your tests.

I've added test with various combinations to assert that this extension works exactly the same as `XCTAssertEqual`.
Coverage is at 100% for this extension. 
<img width="550" alt="Screen Shot 2022-09-15 at 10 55 24 PM" src="https://user-images.githubusercontent.com/15278390/190507477-4bf480a8-8ed6-422d-a2d8-749cfeb151cb.png">

borut-t commented on Tests/Tests/Core/Extensions/Other/XCTestCaseTests.swift:
```diff
@@ -0,0 +1,78 @@
+    // Test passes even though httpBody is different.
+    XCTAssertEqual(firstUrlRequest, secondUrlRequest)
+    
+    XCTExpectFailure("Expected test to fail since httpBody is different.") {
+      XCTAssertEqual(firstUrlRequest.httpBody, secondUrlRequest.httpBody)
+      XCTAssertEqualURLRequest(firstUrlRequest, secondUrlRequest)
+    }
```
Nice use of block-based expectation 👌

borut-t commented on Tests/Tests/Core/Extensions/Other/XCTestCaseTests.swift:
```diff
@@ -0,0 +1,78 @@
+    }
+  }
+  
+  func test_XCTAssertEqualURLRequest_hasExpectedFailures() {
```
Do you prefer snake case over a camel case?

gentique replied:
Hmm, this naming convention is something I got from the Essential Developer course. I like it since it makes it very clear what is being tested and what is being expected from the test without having to go into the test itself.

borut-t replied:
I don't mind about the naming at all. Just the case type.
```swift
func test_XCTAssertEqualURLRequest_hasExpectedFailures()
vs 
func testXCTAssertEqualURLRequestHasExpectedFailures()
```
Again, don't have strong feelings. Maybe we should vote about it 😄

gentique replied:
Sorry, I meant just what you said:

test_`System under test`_`Expected behaviour` the underline makes it easier to read in my opinion 🤷‍♂️ , maybe I just got used to it by now. We could vote on it on our next pod meeting 😄

borut-t replied:
Maybe it's just me that I'm not used to this snake case. But I get your point. It is easier to read. And for tests, it's ok I think 👌 We can still do a quick vote on this at the next pod meeting 😄

borut-t commented on Sources/Core/Extensions/Other/XCTestCase+PovioKit.swift:
```diff
@@ -0,0 +1,20 @@
+import XCTest
+
+public extension XCTestCase {
+  func XCTAssertEqualURLRequest(_ lhs: URLRequest?, _ rhs: URLRequest?, file: StaticString = #filePath, line: UInt = #line) {
```
Do we want to have the `NotEqual` method as well? 🤔

gentique replied:
I'm not sure honestly, maybe we can add it later on if such a need appears.

What do you think?

borut-t replied:
It is useful in some cases. And to be consistent with XCTest framework, maybe we should add it 🤷‍♂️ . I don't have strong feelings about it though.

borut-t reviewed (approved):
Great work @gentique. Added a few nits below. I'll rebase my branch once this gets merged.

borut-t commented:
@gentique just run tests and all your tests are failing 🤔

gentique commented:
> @gentique just run tests and all your tests are failing 🤔

Well yes, the way I have written these tests is to Assert that they correctly fail just like XCTAssert does, that's why they are wrapped in `XCTExpectFailure`. 

in cases where you have different http method, body, headers, you need to assert that the extension fails (this is the expected behaviour).

<img width="1130" alt="Screen Shot 2022-09-16 at 11 41 21 AM" src="https://user-images.githubusercontent.com/15278390/190610625-9892c3ba-8f04-4746-8e34-e34a36bdd6e9.png">
You can see in the above tests that it does not show up as an actual failure.

I will rewrite this since upon making these tests "pass" (which is not expected) they don't report as failing.

borut-t commented:
> > @gentique just run tests and all your tests are failing 🤔
> 
> Well yes, the way I have written these tests is to Assert that they correctly fail just like XCTAssert does, that's why they are wrapped in `XCTExpectFailure`.
> 
> in cases where you have different http method, body, headers, you need to assert that the extension fails (this is the expected behaviour).
> 
> <img alt="Screen Shot 2022-09-16 at 11 41 21 AM" width="1130" src="https://user-images.githubusercontent.com/15278390/190610625-9892c3ba-8f04-4746-8e34-e34a36bdd6e9.png">
> 
> You can see in the above tests that it does not show up as an actual failure. --
> #### I will rewrite this since upon making these tests "pass" (which is not expected) they don't report as failing.

My bad, they succeed indeed.

gentique commented:
- Added `XCTAssertNotEqualURLRequest`
  - Add tests
- Refactored tests
- Add documentation

markom-povio reviewed (approved):
Nice 👍
